File: samples/android/app/src/main/java/app/serenada/sample/SampleAudioCoordinator.kt
package app.serenada.sample

import app.serenada.core.call.AudioDevice
import app.serenada.core.call.AudioDeviceDirection
import app.serenada.core.call.AudioDeviceKind
import app.serenada.core.call.AudioDeviceStatus
import app.serenada.core.call.AudioIntent
import app.serenada.core.call.AudioCoordinatorEvent
import app.serenada.core.call.SerenadaAudioCoordinator
import kotlinx.coroutines.flow.MutableSharedFlow
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.SharedFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asSharedFlow
import kotlinx.coroutines.flow.asStateFlow
import android.util.Log

class SampleAudioCoordinator : SerenadaAudioCoordinator {
    private val TAG = "SampleAudioCoordinator"

    private val speaker = AudioDevice(
        id = "speaker",
        displayName = "Mock Speaker",
        kind = AudioDeviceKind.Speakerphone,
        direction = AudioDeviceDirection.OUTPUT,
        status = AudioDeviceStatus.ACTIVE
    )
    
    private val mic = AudioDevice(
        id = "mic",
        displayName = "Mock Mic",
        kind = AudioDeviceKind.Earpiece,
        direction = AudioDeviceDirection.INPUT,
        status = AudioDeviceStatus.ACTIVE
    )

    private val _availableDevices = MutableStateFlow<List<AudioDevice>>(listOf(speaker, mic))
    override val availableDevices: StateFlow<List<AudioDevice>> = _availableDevices.asStateFlow()

    private val _effectiveInputDevice = MutableStateFlow<AudioDevice?>(mic)
    override val effectiveInputDevice: StateFlow<AudioDevice?> = _effectiveInputDevice.asStateFlow()

    private val _effectiveOutputDevice = MutableStateFlow<AudioDevice?>(speaker)
    override val effectiveOutputDevice: StateFlow<AudioDevice?> = _effectiveOutputDevice.asStateFlow()

    private val _events = MutableSharedFlow<AudioCoordinatorEvent>(extraBufferCapacity = 64)
    override val events: SharedFlow<AudioCoordinatorEvent> = _events.asSharedFlow()

    override suspend fun activateCallSession(intent: AudioIntent) {
        Log.d(TAG, "activateCallSession called with intent: $intent")
    }

    override suspend fun deactivateCallSession() {
        Log.d(TAG, "deactivateCallSession called")
    }

    override suspend fun applyRouting(device: AudioDevice) {
        Log.d(TAG, "applyRouting called for device: ${device.displayName}")
        _effectiveOutputDevice.value = device
        _events.tryEmit(AudioCoordinatorEvent.EffectiveRouteChanged(_effectiveInputDevice.value, device))
    }

    override suspend fun setMicMuted(muted: Boolean) {
        Log.d(TAG, "setMicMuted: $muted")
    }

    fun simulateExternalAudio(active: Boolean) {
        if (active) {
            Log.d(TAG, "Simulating external audio start")
            _events.tryEmit(AudioCoordinatorEvent.ExternalAudioStarted)
        } else {
            Log.d(TAG, "Simulating external audio end")
            _events.tryEmit(AudioCoordinatorEvent.ExternalAudioEnded)
        }
    }
}

File: samples/ios/SampleApp/SampleAudioCoordinator.swift
import AVFoundation
import Foundation
import SerenadaCore

public final class SampleAudioCoordinator: SerenadaAudioCoordinator, @unchecked Sendable {
    private let lock = NSLock()
    private var continuations: [UUID: AsyncStream<[AudioDevice]>.Continuation] = [:]
    private var inputContinuations: [UUID: AsyncStream<AudioDevice?>.Continuation] = [:]
    private var outputContinuations: [UUID: AsyncStream<AudioDevice?>.Continuation] = [:]
    private var eventContinuations: [UUID: AsyncStream<AudioCoordinatorEvent>.Continuation] = [:]

    private var devices: [AudioDevice] = []
    private var activeInput: AudioDevice?
    private var activeOutput: AudioDevice?

    public init() {
        let speaker = AudioDevice(id: "speaker", displayName: "Mock Speaker", kind: .speakerphone, direction: .output, status: .active)
        let mic = AudioDevice(id: "mic", displayName: "Mock Mic", kind: .earpiece, direction: .input, status: .active)
        self.devices = [speaker, mic]
        self.activeInput = mic
        self.activeOutput = speaker
    }

    public func activateCallSession(intent: AudioIntent) async throws {
        print("[SampleAudioCoordinator] activateCallSession called with intent: \(intent)")
        let session = AVAudioSession.sharedInstance()
        try session.setCategory(.playAndRecord, mode: .voiceChat, options: [.allowBluetoothHFP, .allowBluetoothA2DP, .mixWithOthers])
        try session.setActive(true)
    }

    public func deactivateCallSession() async {
        print("[SampleAudioCoordinator] deactivateCallSession called")
        try? AVAudioSession.sharedInstance().setActive(false, options: .notifyOthersOnDeactivation)
    }

    public func applyRouting(_ device: AudioDevice) async throws {
        print("[SampleAudioCoordinator] applyRouting called for device: \(device.displayName)")
        lock.lock()
        activeOutput = device
        let currentOutputContinuations = Array(outputContinuations.values)
        let currentInput = activeInput
        lock.unlock()
        
        for c in currentOutputContinuations {
            c.yield(device)
        }
        
        emit(.effectiveRouteChanged(input: currentInput, output: device))
    }

    public func setMicMuted(_ muted: Bool) async throws {
        print("[SampleAudioCoordinator] setMicMuted: \(muted)")
    }

    public func simulateExternalAudio(_ active: Bool) {
        if active {
            print("[SampleAudioCoordinator] Simulating external audio start")
            emit(.externalAudioStarted)
        } else {
            print("[SampleAudioCoordinator] Simulating external audio end")
            emit(.externalAudioEnded)
        }
    }

    private func emit(_ event: AudioCoordinatorEvent) {
        lock.lock()
        let currentEventContinuations = Array(eventContinuations.values)
        lock.unlock()
        
        for c in currentEventContinuations {
            c.yield(event)
        }
    }

    public var availableDevices: AsyncStream<[AudioDevice]> {
        AsyncStream { continuation in
            let id = UUID()
            lock.lock()
            continuations[id] = continuation
            let initial = devices
            lock.unlock()
            
            continuation.yield(initial)
            
            continuation.onTermination = { [weak self] _ in
                self?.lock.lock()
                self?.continuations.removeValue(forKey: id)
                self?.lock.unlock()
            }
        }
    }

    public var effectiveInputDevice: AsyncStream<AudioDevice?> {
        AsyncStream { continuation in
            let id = UUID()
            lock.lock()
            inputContinuations[id] = continuation
            let initial = activeInput
            lock.unlock()
            
            continuation.yield(initial)
            
            continuation.onTermination = { [weak self] _ in
                self?.lock.lock()
                self?.inputContinuations.removeValue(forKey: id)
                self?.lock.unlock()
            }
        }
    }

    public var effectiveOutputDevice: AsyncStream<AudioDevice?> {
        AsyncStream { continuation in
            let id = UUID()
            lock.lock()
            outputContinuations[id] = continuation
            let initial = activeOutput
            lock.unlock()
            
            continuation.yield(initial)
            
            continuation.onTermination = { [weak self] _ in
                self?.lock.lock()
                self?.outputContinuations.removeValue(forKey: id)
                self?.lock.unlock()
            }
        }
    }

    public var events: AsyncStream<AudioCoordinatorEvent> {
        AsyncStream { continuation in
            let id = UUID()
            lock.lock()
            eventContinuations[id] = continuation
            lock.unlock()
            
            continuation.onTermination = { [weak self] _ in
                self?.lock.lock()
                self?.eventContinuations.removeValue(forKey: id)
                self?.lock.unlock()
            }
        }
    }
}
